android-pay/src/main/java/com/stripe/wrap/pay/utils/CartManager.java

```diff
@@ -0,0 +1,122 @@
+    private List<LineItem> mLineItemsShipping = new ArrayList<>();
+    private LineItem mLineItemTax;
+
+    public CartManager() {
+        mCurrency = Currency.getInstance(Locale.getDefault());
+    }
+
+    public CartManager(String currencyCode) {
+        setCurrencyCode(currencyCode);
+    }
+
+    /**
+     * Sets the ISO 4217 currency code of the line item. If the input currency is invalid,
+     * currency is set to the default for the phone's locale.
+     *
+     * @param currencyCode the currency code to set
+     * @return {@code this}, for chaining purposes
+     */
+    public CartManager setCurrencyCode(String currencyCode) {
+        mCurrency = PaymentUtils.getCurrencyByCodeOrDefault(currencyCode);
+        return this;
+    }
+
+    /**
+     * Add a {@link LineItem} to the cart.
+     *
+     * @param item the {@link LineItem} to be added
+     */
+    public CartManager addLineItem(@NonNull LineItem item) {
+        switch (item.getRole()) {
+            case LineItem.Role.REGULAR:
+                mLineItemsRegular.add(item);
+                break;
+            case LineItem.Role.SHIPPING:
+                mLineItemsShipping.add(item);
+                break;
+            case LineItem.Role.TAX:
+                if (mLineItemTax != null) {
+                    Log.w(TAG, String.format(Locale.ENGLISH,
+                            "Adding a tax line item, but a tax line item " +
+                            "already exists. Old tax of %s is being overwritten " +
+                            "to maintain a valid cart.",
+                            mLineItemTax.getTotalPrice()));
+                }
+                mLineItemTax = item;
+                break;
+            default:
+                Log.w(TAG, String.format(Locale.ENGLISH,
+                        "Line item with unknown role added to cart. Treated as regular. " +
+                        "Unknown role is of code %d",
+                        item.getRole()));
+                mLineItemsRegular.add(item);
+                break;
+        }
+        return this;
+    }
+
+    /**
+     * Build the {@link Cart}. Returns {@code null} if the item set is invalid.
+     *
+     * @return a {@link Cart}, or {@code null} if any of the items are invalid
+     */
+    @Nullable
+    public Cart build() {
+        List<LineItem> totalLineItems = new ArrayList<>();
+        totalLineItems.addAll(mLineItemsRegular);
+        totalLineItems.addAll(mLineItemsShipping);
+        totalLineItems.add(mLineItemTax);
+
+        if (PaymentUtils.isLineItemListValid(
+                totalLineItems,
+                mCurrency.getCurrencyCode())) {
+            return Cart.newBuilder()
+                    .setCurrencyCode(mCurrency.getCurrencyCode())
+                    .setLineItems(totalLineItems)
+                    .setTotalPrice(PaymentUtils.getTotalPriceString(totalLineItems, mCurrency))
+                    .build();
+        } else {
+            return null;
```

Comment:
This is the function that's going to change -- I'm going to have them pass in an error-output list, so that they can see which LineItems are causing the error. (I will also return null)